Dayview category feature

KreivenWang opened the pull request:
based on branch [separate-view-and-model](https://github.com/KreivenWang/Ionic2-Calendar/tree/separate-view-and-model)

KreivenWang committed:
implement dayview category

KreivenWang committed:
loop slider content in weekview to deduplicate view code

KreivenWang committed:
Update README.md

KreivenWang committed:
Update README.md with customized category css and template

KreivenWang committed:
Update README.md

KreivenWang committed:
update package version

KreivenWang committed:
fix layout - align category header border to content border

KreivenWang committed:
update package version

KreivenWang committed:
update readme and package version

KreivenWang committed:
fix dayview scrollToHour

twinssbc commented on src/weekview.ts:
```diff
@@ -544,6 +88,7 @@ export class WeekViewComponent implements ICalendarComponent, OnInit, OnChanges,
     @Output() onDayHeaderSelected = new EventEmitter<ITimeSelected>();
     @Output() onTitleChanged = new EventEmitter<string>();
 
+    public readonly sliderIndexList = [0, 1, 2];
```
where is this used?

twinssbc commented on README.md:
add category related info in the DataSource section, describe how to pass the field so that day view can use it proplery.

twinssbc commented on src/dayview.ts:
```diff
@@ -841,23 +596,121 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
+                });
             }
         }
+        return Array.from(categoryIdCategoryMap.values());
```
just use Set, so that later the find method will be more efficient.

twinssbc commented on src/dayview.ts:
```diff
@@ -741,59 +452,99 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
             }
 
+            const categoryId = event.categoryId || this.getDefaultCategoryId();
+            const category = allCategories.find(
```
is allCategories a Map? If it's an array, the find is O(N) which is slow

twinssbc commented on src/dayview.ts:
```diff
@@ -841,23 +596,121 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
+        console.log('categorizedAllDayEvents', categorizedAllDayEvents);
+    }
+
+    getDefaultCategoryId(): string {
```
private

twinssbc commented on src/dayview.ts:
```diff
@@ -841,23 +596,121 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
+                        displayEvent.event.categoryId ||
+                        this.getDefaultCategoryId();
+                    if (groupedEvents) {
+                        if (Array.isArray(groupedEvents.get(categoryId))) {
```
will this ever be an array?

twinssbc commented on src/dayview.ts:
```diff
@@ -741,59 +453,99 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
             eps = 0.016,
             rangeStartRowIndex = this.startHour * this.hourSegments,
             rangeEndRowIndex = this.endHour * this.hourSegments;
+        const allCategories = this.extractAllCategories(eventSource);
```
this extractAllCategories is a heavy operation as need to loop the whole eventSource. if should be only used when dayviewShowCategoryView is true.
also add the option that the user could directly passed the category as an option, as user should have such knowlege. If the cateory is already passed, then no need to extract from eventSource.

twinssbc commented on src/dayview.ts:
```diff
@@ -841,23 +597,121 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
                         rows[startIndex].events = eventSet;
                     }
+
+                    // setup eventsGroupedByCategory
```
only do this when dayviewShowCategoryView is true

KreivenWang replied:
fixed

twinssbc commented on src/dayview.ts:
```diff
@@ -841,23 +597,121 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
+            } else {
+                categories.unshift(defaultCategory);
+            }
+            console.log("placed default category: ", categories);
```
dont log

twinssbc commented on src/dayview.ts:
```diff
@@ -841,23 +597,121 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
+        categoryIdAllDayEventsMap.forEach((events) => {
+            categorizedAllDayEvents.push(events.map((event) => ({ event })));
+        });
+        console.log("categorizedAllDayEvents", categorizedAllDayEvents);
```
dont log

KreivenWang committed:
Revert "fix dayview scrollToHour"

This reverts commit 3fed1e0dd2fc8eb61d5bfb2073f43abfd4a09882.

KreivenWang committed:
update by pr comment - use category map instead of array

KreivenWang committed:
Merge branch 'separate-view-and-model' into dayview-category

KreivenWang committed:
user category source instead of implying from event source

KreivenWang committed:
setup display categories

KreivenWang committed:
correct map of defaultCategoryItemTemplate

KreivenWang committed:
clean up - remove DefaultCategoryPlacement; line up formatter

KreivenWang committed:
Update README.md

KreivenWang committed:
Update README.md

KreivenWang committed:
Update README.md

KreivenWang committed:
set dayviewCategorySource optional without default new Set

KreivenWang committed:
Merge remote-tracking branch 'origin/ionic7' into dayview-category

KreivenWang committed:
revert package.json to remove prettier config

KreivenWang committed:
fixed a bug where dayview normalEventInRange is not called placeEvents

KreivenWang committed:
format dayview.ts more back to original format

KreivenWang committed:
format dayview.ts more back to original format

KreivenWang committed:
fixed a bug where dayview normalEventInRange is not called placeEvents

twinssbc commented on src/dayview.html:
```diff
@@ -1,65 +1,167 @@
-                                </ng-template>
+                      <tr>
+                        <ng-container *ngIf="sliderIndex === currentViewIndex">
+                          <ng-container *ngFor="let allDayEventsCategory of views[sliderIndex].categorizedAllDayEvents">
```
use categorizedAllDayEventsMap, then you don't need this for loop, simply fetch category as the key, in td directly use views[sliderIndex].categorizedAllDayEventsMap[category]

twinssbc commented on src/dayview.html:
```diff
@@ -1,65 +1,167 @@
+                      [ngTemplateOutletContext]="{allDayEvents:views[sliderIndex].allDayEvents,eventTemplate:dayviewAllDayEventTemplate}">
+                    </ng-template>
+                  </td>
+                  <td class="calendar-cell" *ngIf="sliderIndex!==currentViewIndex">
```
for the inactive view, should have the same layout as active view

twinssbc commented on src/dayview.html:
```diff
@@ -1,65 +1,167 @@
+            </thead>
+          </table>
+        </div>
+        <div *ngIf="sliderIndex!==currentViewIndex" class="dayview-inactive-category-header"></div>
```
for the inactive view, need to have the same layout as active view.

KreivenWang replied:
resolved

twinssbc commented on src/dayview.html:
```diff
@@ -1,65 +1,167 @@
+              <tr>
+                <th class="dayview-category-header-item calendar-hour-column"></th>
+                <th *ngFor="let category of views[sliderIndex].categories" class="dayview-category-header-item">
+                  <div class="dayview-category-name">
```
no need this extra div, just apply the style on dayview-catetory-header-item

twinssbc commented on src/dayview.html:
```diff
@@ -1,65 +1,167 @@
-                        </tr>
+                          </ng-container>
+                        </ng-container>
+                        <td class="calendar-cell" *ngIf="sliderIndex !== currentViewIndex">
```
need to use the same layout on inactive view

twinssbc commented on src/dayview.html:
```diff
@@ -1,65 +1,167 @@
+                <td *ngFor="let category of views[sliderIndex].categories" class="calendar-cell"
+                  (click)="select(row, category)" tappable>
+                  <ng-container *ngFor="let categorizedEvents of row.eventsGroupByCategory | keyvalue">
+                    <ng-template *ngIf="categorizedEvents.key === category"
```
no need this extra for loop, just use rows.eventsGroupByCategory[category]

twinssbc commented on src/dayview.html:
```diff
@@ -1,65 +1,167 @@
+                <td class="calendar-hour-column text-center">
+                  {{ hourColumnLabels[i] }}
+                </td>
+                <td *ngIf="views[sliderIndex].categories.length === 0" class="calendar-cell" (click)="select(row)"
```
is this necessary, just assume if need to show categorized view, always have category number larger than 0

KreivenWang replied:
yes it's necessary, it's used as a placeholder but only there is no event. Remove it then the hour column 'td' will be the only element which occupies whole row like this:

KreivenWang replied:
![image](https://github.com/twinssbc/Ionic2-Calendar/assets/27603495/f33986ef-fede-4c8d-ac6f-9d64a64b5710)

twinssbc replied:
why not using dayviewCategorySource to generate all td? it is the predefined static list of all category names

twinssbc commented on src/dayview.html:
in general, 
1. need to apply the same HTML layout on inactive view as active view.
2. use Map to group the categorized event, key is category, then directly fetch the values directly, instead of using a for loop then use if to locate the matched category.

twinssbc reviewed (changes requested):
in general,
1. need to apply the same HTML layout on inactive view as active view.
2. use Map to group the categorized event, key is category, then directly fetch the values directly, instead of using a for loop then use if to locate the matched category.

KreivenWang committed:
replace ngFor with map.get() in html template of displaying categorizedAllDayEvents

KreivenWang committed:
update by comment: no need extra div, just apply the style on dayview-catetory-header-item

KreivenWang committed:
for the inactive view, have the same layout as active view

KreivenWang committed:
Merge remote-tracking branch 'origin/ionic7' into dayview-category

KreivenWang committed:
for the inactive view, have the same layout as active view

KreivenWang committed:
use map.get(category) instead of extra loop

KreivenWang committed:
remove unnecessary view when categories.length === 0

KreivenWang commented:
updated as following
1. have aligned inactive view with active view
2. removed unnecessary loop from category map
3. checked and fixed other comments one by one

KreivenWang commented:
4. merged from latest ionic7 branch and resolved conflict for swiper11 upgrade

KreivenWang committed:
Revert "remove unnecessary view when categories.length === 0"

This reverts commit 35c1545bfa397be54b6253b9ce3b63966b5148ca.

KreivenWang committed:
use map.get(category) instead of extra loop

KreivenWang committed:
fix a bug where weekview loop misplaced

KreivenWang committed:
fix a css conflicts caused by none view encapsulation between weekview and dayview table styles. Issue happened in weekview that header showed at bottom in chrome after switching from dayview

twinssbc commented on README.md:
```diff
@@ -148,6 +150,14 @@ export class MyPageModule {}
 
 # Options
 
+* `dayviewShowCategoryView` (version: 2.2+)    
```
intend to bump to version 2.4 after merge, update the README to 2.4

twinssbc commented on src/dayview.html:
```diff
@@ -1,187 +1,181 @@
+        <div class="dayview-allday-table">
+          <div class="dayview-allday-label">{{allDayLabel}}</div>
+          <div class="dayview-allday-content-wrapper scroll-content">
+            <table class="dayview-table dayview-table-bordered dayview-allday-content-table">
```
still add table in the class, as user on previous version may already customize based on ".table"

twinssbc replied:
you could use '.dayview-slide-container table' to add customized style

twinssbc commented on src/dayview.html:
```diff
@@ -1,187 +1,181 @@
+        <init-position-scroll *ngIf="sliderIndex===currentViewIndex" class="dayview-normal-event-container"
+          [initPosition]="initScrollPosition" [emitEvent]="preserveScrollPosition"
+          (onScroll)="setScrollPosition($event)">
+          <table class="dayview-table dayview-table-bordered dayview-table-fixed dayview-normal-event-table">
```
still add table table-bordered table-fixed class

twinssbc commented on src/dayview.html:
```diff
@@ -1,187 +1,181 @@
+        </init-position-scroll>
+        <init-position-scroll *ngIf="sliderIndex!==currentViewIndex" class="dayview-normal-event-container"
+          [initPosition]="initScrollPosition">
+          <table class="dayview-table dayview-table-bordered dayview-table-fixed dayview-normal-event-table">
```
still add table table-bordered table-fixed class

twinssbc commented on src/dayview.ts:
```diff
@@ -424,19 +443,23 @@ export class DayViewComponent implements ICalendarComponent, OnInit, OnChanges,
                     event
                 });
+                if (event.category && this.dayviewCategorySource && this.dayviewCategorySource.has(event.category)) {
+                    displayCategorySet.add(event.category);
```
do you still need displaycategorySet? dayviewcategorySource controls all the events should be displayed?

KreivenWang replied:
`displaycategorySet` is still required to display categories only for current day. If we use `dayviewcategorySource`, all categories will display, there will be many empty event columns from other days.

twinssbc commented on src/weekview.html:
```diff
@@ -1,11 +1,11 @@
-        <div class="swiper-slide slide-container weekview-slide-container">
-            <table class="table table-bordered table-fixed weekview-header">
+        <div *ngFor="let sliderIndex of sliderIndexList" class="swiper-slide slide-container weekview-slide-container">
+            <table class="weekview-table weekview-table-bordered weekview-table-fixed weekview-header">
```
still add table table-bordered table-fixed, use "weekview-slide-container > table" to customize.

KreivenWang committed:
update readme to tag new feature with 2.4+

KreivenWang committed:
remove usage of displaycategorySet; replace with dayviewcategorySource

KreivenWang committed:
Revert "fix a css conflicts caused by none view encapsulation between weekview and dayview table styles. Issue happened in weekview that header showed at bottom in chrome after switching from dayview"

KreivenWang committed:
use container css to resolve css conflict instead of rename

KreivenWang committed:
use container css to resolve css conflict instead of rename

KreivenWang committed:
Revert "remove usage of displaycategorySet; replace with dayviewcategorySource" it's still required to filter categories to display for current day

KreivenWang committed:
Revert "Revert "remove usage of displaycategorySet; replace with dayviewcategorySource" it's still required to filter categories to display for current day"

This reverts commit 05d3e787d394d6b6cb1ed0d051786e7d4bdf70bd.

KreivenWang committed:
fix inactive dayview layout

twinssbc commented on src/dayview.html:
```diff
@@ -1,187 +1,181 @@
+            <table class="table table-fixed dayview-allday-category-table">
+              <tbody>
+                <td class="dayview-allday-category-cell" *ngFor="let category of dayviewCategorySource">
+                  <table class="table dayview-allday-content-table">
```
I don't think it's necessary to nest a table inside dayview-allday-category-table, it seems just a one cell table.

twinssbc replied:
I don't think it's necessary to nest a table inside dayview-allday-category-table, it seems just a one cell table.

twinssbc commented on src/dayview.html:
```diff
@@ -1,187 +1,181 @@
+      </ng-container>
+
+      <ng-container *ngIf="dayviewShowCategoryView">
+        <div *ngIf="sliderIndex===currentViewIndex" class="dayview-category-header">
```
line 68-80 and line 81-93 seem duplicated.

twinssbc commented on src/dayview.html:
```diff
@@ -1,187 +1,181 @@
+                </td>
+                <td *ngFor="let category of dayviewCategorySource" class="calendar-cell"
+                  (click)="select(row, category)" tappable>
+                  <ng-template *ngIf="row.eventsGroupByCategory?.get(category)?.length"
```
I think even current category doesn't have event, you still want to display the style for dayviewNormalEventSectionTemplate, so should not use *ngIf to exclude it. Otherwise the cells having event and not having event may have different style.

twinssbc commented on src/dayview.html:
```diff
@@ -1,187 +1,181 @@
+                </td>
+                <td *ngFor="let category of dayviewCategorySource" class="calendar-cell"
+                  (click)="select(row, category)" tappable>
+                  <ng-template *ngIf="row.eventsGroupByCategory?.get(category)?.length"
```
I think even current category doesn't have event, you still want to display the style for dayviewInactiveNormalEventSectionTemplate, so should not use *ngIf to exclude it. Otherwise the cells having event and not having event may have different style.

KreivenWang committed:
enhance dayview html based on PR comments

twinssbc commented on src/dayview.css:
```diff
@@ -87,14 +87,36 @@
+  overflow: hidden;
+}
+
+.dayview-category-cell-container {
```
no used, remove it

twinssbc commented on src/dayview.css:
```diff
@@ -87,14 +87,36 @@
+  display: flex;
+}
+
+.dayview-category-cell-item {
```
not used, remove it, also remove it in README

twinssbc commented on src/dayview.css:
```diff
@@ -87,14 +87,36 @@
+.dayview-category-header-item {
+  font-weight: normal;
+  height: 37px;
+  border-left: 1px solid #ddd;
```
have you tested if the border-left works in RTL mode? maybe you don't need it as you have table-bordered specified?

twinssbc commented on src/dayview.css:
```diff
@@ -113,13 +135,21 @@
+}
+
+.dayview-allday-category-cell {
+  border: solid 1px #ddd;
```
no need, as you have table-bordered at the parent level

twinssbc commented on src/dayview.css:
```diff
@@ -113,13 +135,21 @@
-  border-left: 1px solid #ddd;
-  border-right: 1px solid #ddd;
+.dayview-allday-category-table {
+  border: solid 1px #ddd;
```
no need, as you have table-bordered at the parent level

twinssbc commented on src/dayview.css:
```diff
@@ -113,13 +135,21 @@
 
-.dayview-allday-content-table {
-  min-height: 50px;
+.dayview-allday-category-wrapper {
```
not sure you still need this, as we already have style defined as below

.dayview-allday-content-wrapper {
  margin-left: 50px;
  overflow: hidden;
  height: 51px;
}

KreivenWang replied:
it's used to override height part of dayview-allday-content-wrapper. The new wrapper shifts the height 1px up so that the borders of allday content table can overlap with existing borders of above category table and below content table.
Without the override, the top and bottom borders look like duplicated:
![image](https://github.com/twinssbc/Ionic2-Calendar/assets/27603495/e28cb121-627f-4956-957f-e7f6c368800f)

KreivenWang committed:
clean up unused dayview css

KreivenWang committed:
clean up unused borders from dayview css

KreivenWang committed:
rtl test fix

KreivenWang commented:
Have also tested in RTL mode

twinssbc commented on src/dayview.css:
```diff
@@ -1,58 +1,58 @@
+.dayview-slide-container .table {
   width: 100%;
   max-width: 100%;
+  height: 100%;
```
remove this line? it was not there

KreivenWang committed:
revert dayview table container height; fix category allday table height
